be/dtm0_log.c

```diff
@@ -0,0 +1,631 @@
+			M0_BE_ALLOC_BUF_SYNC(lpayload, seg, tx);
+			m0_buf_memcpy(lpayload, payload);
+			M0_BE_TX_CAPTURE_BUF(seg, tx, lpayload);
+			M0_BE_TX_CAPTURE_PTR(seg, tx, lpayload);
+		} else {
+			m0_buf_copy(lpayload, payload);
+		}
+	}
+
+	m0_dtm0_tx_desc_apply(&rec->dlr_txd, txd);
+	M0_POST(m0_dtm0_log_rec__invariant(rec));
+	if (is_persistent) {
+		M0_BE_TX_CAPTURE_ARR(seg, tx, rec->dlr_txd.dtd_ps.dtp_pa,
+					      rec->dlr_txd.dtd_ps.dtp_nr);
+	}
+
+	return 0;
+}
+
+M0_INTERNAL int m0_be_dtm0_log_update(struct m0_be_dtm0_log  *log,
+                                      struct m0_be_tx        *tx,
+                                      struct m0_dtm0_tx_desc *txd,
+                                      struct m0_buf          *payload)
+{
+	struct m0_dtm0_log_rec	*rec;
+
+	M0_PRE(payload != NULL);
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tx_desc__invariant(txd));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	return (rec = m0_be_dtm0_log_find(log, &txd->dtd_id)) ?
+                be_dtm0_log__set(log, tx, txd, payload, rec) :
+                m0_be_dtm0_log__insert(log, tx, txd, payload);
+}
+
+M0_INTERNAL int m0_be_dtm0_log_prune(struct m0_be_dtm0_log    *log,
+                                     struct m0_be_tx          *tx,
+                                     const struct m0_dtm0_tid *id)
+{
+	/* This assignment is meaningful as it covers the empty log case */
+	int                     rc = M0_DTS_LT;
+	struct m0_dtm0_log_rec *rec;
+	struct m0_dtm0_log_rec *currec;
+
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tid__invariant(id));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	m0_tl_for (lrec, log->dl_tlist, rec) {
+		if (!m0_dtm0_tx_desc_state_eq(&rec->dlr_txd,
+					      M0_DTPS_PERSISTENT))
+			return M0_ERR(-EPROTO);
+
+		rc = m0_dtm0_tid_cmp(log->dl_cs, &rec->dlr_txd.dtd_id, id);
+		if (rc != M0_DTS_LT)
+			break;
+	} m0_tl_endfor;
+
+	if (rc != M0_DTS_EQ)
+		return -ENOENT;
+
+	while ((currec = lrec_tlist_pop(log->dl_tlist)) != rec) {
+		M0_ASSERT(m0_dtm0_log_rec__invariant(currec));
+		be_dtm0_log_rec_fini(&currec, tx);
+	}
+
+	be_dtm0_log_rec_fini(&currec, tx);
+	return rc;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_clear(struct m0_be_dtm0_log *log)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	/* This function is expected to be called only on the client side where
+	   the log will always be a volatile log. */
+	M0_ASSERT(!log->dl_is_persistent);
+
+	m0_tl_teardown(lrec, log->dl_tlist, rec) {
+		M0_ASSERT(m0_dtm0_log_rec__invariant(rec));
+		M0_ASSERT(m0_dtm0_tx_desc_state_eq(&rec->dlr_dtx.dd_txd,
+						   M0_DTPS_PERSISTENT));
+		be_dtm0_log_rec_fini(&rec, NULL);
+	}
+	M0_POST(lrec_tlist_is_empty(log->dl_tlist));
+}
+
+M0_INTERNAL int m0_be_dtm0_volatile_log_insert(struct m0_be_dtm0_log  *log,
+					       struct m0_dtm0_log_rec *rec)
+{
+	int rc;
+
+	/* TODO: dissolve dlr_txd and remove this code */
+	rc = m0_dtm0_tx_desc_copy(&rec->dlr_dtx.dd_txd, &rec->dlr_txd);
+	if (rc != 0)
+		return rc;
```

Comment:
Add M0_ENTRY() at the beginning of a function, and
return M0_RC(rc);
at the end of a function.

Here, we can use:
return M0_ERR(rc);
here.

Reply:
Done.